### PR TITLE
feat: disable year and month navigation if min date specified

### PR DESCRIPTION
**Pull request by pitama**

### CODE REVIEW

**Comment by wesleycho** [comment]
Can you provide a [Plunker](http://plnkr.co/) with this in action?

**Comment by pitama** [comment]
I found som bug, I fix it. Here is the plunker http://plnkr.co/edit/q4fuEcFqGSgzs85RsblW?p=preview

File: src/datepicker/datepicker.js [hunk 1]
```diff
@@ -268,6 +268,26 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
           });
         }
 
+        // disable navigation according to the range (maxDate and minDate)
+        // compare minDate with the first day of the month
+        var calendarMinDate = new Date(firstDayOfMonth.getTime());
+        calendarMinDate.setHours(0);
+        // if the first day of the month is not greater then minDate there is no point to navigate to a previous month
+        if (ctrl.minDate != null) {
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Is the intention to catch other values than null or also to catch undefined and other falsy values? An explicit check for the desired statement is preferred here for readability

File: src/datepicker/datepicker.js [hunk 2]
```diff
@@ -268,6 +268,26 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
...
+          scope.navPrevDisabled = ctrl.minDate && (calendarMinDate <= ctrl.minDate);
+        } else {
+          scope.navPrevDisabled = false;
+        }
+        // compare maxDate with the first day of the next month
+        var calendarMaxDate = new Date(firstDayOfMonth.getTime());
+        calendarMaxDate.setHours(0);
+        calendarMaxDate.setMonth(calendarMaxDate.getMonth() + 1);
+        // if the first day of the next month is greater then maxDate there is no point to navigate to the next month
+        if (ctrl.maxDate != null) {
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Ditto

File: src/datepicker/datepicker.js [hunk 3]
```diff
@@ -351,6 +371,12 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
           });
         }
 
+        // disable navigation according to the range (maxDate and minDate)
+        // if the first day of the year is not greater then minDate there is no point to navigate to a previous year
+        scope.navPrevDisabled = ctrl.minDate && (months[0].date <= ctrl.minDate);
+        // if the first day of the next year is greater then maxDate there is no point to navigate to the next year
+        scope.navNextDisabled = ctrl.maxDate && (new Date(year+1, 0, 1) > ctrl.maxDate);
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Add spacing around `+`

File: src/datepicker/datepicker.js [hunk 4]
```diff
@@ -411,6 +437,12 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
           });
         }
 
+        // disable navigation according to the range
+        // if the first day of the year range is not greater then minDate there is no point to navigate to a previous year range
+        scope.navPrevDisabled = ctrl.minDate && (years[0].date <= ctrl.minDate);
+        // if the first day of the next year range is greater then maxDate there is no point to navigate to the next year range
+        scope.navNextDisabled = ctrl.maxDate && (new Date(years[range-1].date.getFullYear()+1, 0, 1) > ctrl.maxDate);
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Ditto

File: src/datepicker/test/datepicker.spec.js [hunk 5]
```diff
@@ -802,12 +810,9 @@ describe('datepicker directive', function () {
       expect(element.hasClass('ng-invalid-date-disabled')).toBeTruthy();
     });
 
-    it('disables all days in previous month', function() {
-      clickPreviousButton();
-      var buttons = getAllOptionsEl();
-      angular.forEach(buttons, function( button, index ) {
-        expect(angular.element(button).prop('disabled')).toBe( true );
-      });
+    it('disables the navigation to the previous month', function() {
```
Comment (wesleycho) [inline src/datepicker/test/datepicker.spec.js]: Change to `disables navigation to previous month`

File: src/datepicker/test/datepicker.spec.js [hunk 6]
```diff
@@ -826,13 +831,30 @@ describe('datepicker directive', function () {
       });
     });
 
-    it('disables all months in previous year', function() {
+    it('disables navigation to the previous year', function() {
```
Comment (wesleycho) [inline src/datepicker/test/datepicker.spec.js]: Remove `the`

File: template/datepicker/day.html [hunk 7]
```diff
@@ -1,9 +1,9 @@
 <table role="grid" aria-labelledby="{{::uniqueId}}-title" aria-activedescendant="{{activeDateId}}">
   <thead>
     <tr>
-      <th><button type="button" class="btn btn-default btn-sm pull-left" ng-click="move(-1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-left"></i></button></th>
+      <th><button type="button" class="btn btn-default btn-sm pull-left" ng-disabled="navPrevDisabled" ng-click="move(-1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-left"></i></button></th>
```
Comment (wesleycho) [inline template/datepicker/day.html]: Add bind once to ng-disabled

File: template/datepicker/day.html [hunk 8]
```diff
@@ -1,9 +1,9 @@
...
       <th colspan="{{::5 + showWeeks}}"><button id="{{::uniqueId}}-title" role="heading" aria-live="assertive" aria-atomic="true" type="button" class="btn btn-default btn-sm" ng-click="toggleMode()" ng-disabled="datepickerMode === maxMode" tabindex="-1" style="width:100%;"><strong>{{title}}</strong></button></th>
-      <th><button type="button" class="btn btn-default btn-sm pull-right" ng-click="move(1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-right"></i></button></th>
+      <th><button type="button" class="btn btn-default btn-sm pull-right" ng-disabled="navNextDisabled" ng-click="move(1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-right"></i></button></th>
```
Comment (wesleycho) [inline template/datepicker/day.html]: Ditto

File: template/datepicker/month.html [hunk 9]
```diff
@@ -1,9 +1,9 @@
 <table role="grid" aria-labelledby="{{::uniqueId}}-title" aria-activedescendant="{{activeDateId}}">
   <thead>
     <tr>
-      <th><button type="button" class="btn btn-default btn-sm pull-left" ng-click="move(-1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-left"></i></button></th>
+      <th><button type="button" class="btn btn-default btn-sm pull-left" ng-disabled="navPrevDisabled" ng-click="move(-1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-left"></i></button></th>
```
Comment (wesleycho) [inline template/datepicker/month.html]: Ditto

File: template/datepicker/month.html [hunk 10]
```diff
@@ -1,9 +1,9 @@
...
       <th><button id="{{::uniqueId}}-title" role="heading" aria-live="assertive" aria-atomic="true" type="button" class="btn btn-default btn-sm" ng-click="toggleMode()" ng-disabled="datepickerMode === maxMode" tabindex="-1" style="width:100%;"><strong>{{title}}</strong></button></th>
-      <th><button type="button" class="btn btn-default btn-sm pull-right" ng-click="move(1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-right"></i></button></th>
+      <th><button type="button" class="btn btn-default btn-sm pull-right" ng-disabled="navNextDisabled" ng-click="move(1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-right"></i></button></th>
```
Comment (wesleycho) [inline template/datepicker/month.html]: Ditto

File: template/datepicker/year.html [hunk 11]
```diff
@@ -1,9 +1,9 @@
 <table role="grid" aria-labelledby="{{::uniqueId}}-title" aria-activedescendant="{{activeDateId}}">
   <thead>
     <tr>
-      <th><button type="button" class="btn btn-default btn-sm pull-left" ng-click="move(-1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-left"></i></button></th>
+      <th><button type="button" class="btn btn-default btn-sm pull-left" ng-disabled="navPrevDisabled" ng-click="move(-1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-left"></i></button></th>
```
Comment (wesleycho) [inline template/datepicker/year.html]: Ditto

File: template/datepicker/year.html [hunk 12]
```diff
@@ -1,9 +1,9 @@
...
       <th colspan="3"><button id="{{::uniqueId}}-title" role="heading" aria-live="assertive" aria-atomic="true" type="button" class="btn btn-default btn-sm" ng-click="toggleMode()" ng-disabled="datepickerMode === maxMode" tabindex="-1" style="width:100%;"><strong>{{title}}</strong></button></th>
-      <th><button type="button" class="btn btn-default btn-sm pull-right" ng-click="move(1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-right"></i></button></th>
+      <th><button type="button" class="btn btn-default btn-sm pull-right" ng-disabled="navNextDisabled" ng-click="move(1)" tabindex="-1"><i class="glyphicon glyphicon-chevron-right"></i></button></th>
```
Comment (wesleycho) [inline template/datepicker/year.html]: Ditto

**Comment by wesleycho** [comment]
Overall, I like this PR, although I would like to have a second opinion here.

**Comment by RobJacobs** [comment]
I'm not sure if we want to disable viewing dates outside of the enabled date range. Users may fond it useful to lookup dates/days even though they can't select them.

**Comment by wesleycho** [comment]
Closing as outdated - feel free to file a new PR implementing this feature on top of current `master` if desired.

**Pull request closed without merging**